File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.py
from collections import defaultdict
import heapq


def dijkstra(graph, start, n):
    distance: dict[int, int] = {i: float('inf') for i in range(1, n+1)}
    distance[start] = 0
    
    queue: list[tuple[int, int]] = []
    heapq.heappush(queue, (distance[start], start))
    
    while queue:
        dist, node = heapq.heappop(queue)
        if distance[node] < dist:
            continue
        for next_dist, next_node in graph[node]:
            next_dist += dist
            if distance[next_node] > next_dist:
                distance[next_node] = next_dist
                heapq.heappush(queue, (next_dist, next_node))
    return distance


# 지점 개수, 출발, a, b, [c, d, f원]
def solution(n, s, a, b, fares):
    answer = float('inf')
    
    graph = defaultdict(list)
    for start, end, dist in fares:
        graph[start].append((dist, end))
        graph[end].append((dist, start))
    
    distance_from_s = dijkstra(graph, s, n)
    distance_from_a = dijkstra(graph, a, n)
    distance_from_b = dijkstra(graph, b, n)
    
    for i in range(1, n+1):
        answer = min(
            answer,
            distance_from_s[i] + distance_from_a[i] + distance_from_b[i]
        )
    
    return answer

File: 프로그래머스/3/72413. 합승 택시 요금/합승 택시 요금.java
import java.util.*;

class Solution {
    static List<List<Node>> graph;
    static int INF = Integer.MAX_VALUE;
    
    class Node implements Comparable<Node> {
        int n;
        int w;
        
        Node(int n, int w) {
            this.n = n;
            this.w = w;
        }
        
        @Override
        public int compareTo(Node node) {
            return this.w - node.w;
        }
    }
    
    public int solution(int n, int s, int a, int b, int[][] fares) {
        graph = new ArrayList<>();
        for (int i = 0; i <= n; i++)
            graph.add(new ArrayList<>());
        
        for (int[] fare : fares) {
            int n1 = fare[0], n2 = fare[1], w = fare[2];
            graph.get(n1).add(new Node(n2, w));
            graph.get(n2).add(new Node(n1, w));
        }
        
        int[] dist1 = dijkstra(s, n);
        int[] dist2 = dijkstra(a, n);
        int[] dist3 = dijkstra(b, n);
        
        int answer = INF;
        for (int i = 1; i <= n; i++) {
            answer = Math.min(answer, dist1[i] + dist2[i] + dist3[i]);
        }
        return answer;
    }
    
    private int[] dijkstra(int start, int n) {
        int[] dist = new int[n+1];
        Arrays.fill(dist, INF);
        dist[start] = 0;
        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.offer(new Node(start, 0));
        while (!pq.isEmpty()) {
            Node node = pq.poll();
            if (node.w > dist[node.n]) continue;
            for (Node nextNode : graph.get(node.n)) {
                int cost = node.w + nextNode.w;
                if (dist[nextNode.n] > cost) {
                    dist[nextNode.n] = cost;
                    pq.offer(new Node(nextNode.n, cost));
                }
            }
        }
        return dist;
    }
}